File: php/src/Domain/Purchase/Invoice.php
<?php

namespace Katalogs\BookInvoicing\Domain\Purchase;

use Katalogs\BookInvoicing\Domain\Country\Country;

class Invoice
{
    private array $purchases = [];

    public function __construct(
        public readonly string $customerName,
        public readonly Country $country
    ) {
    }

    public function addPurchase(PurchasedBook $purchase): void
    {
        $this->purchases[] = $purchase;
    }

    public function computeTotalAmount(): float
    {
        $sum = 0;
        /** @var PurchasedBook $purchase */
        foreach ($this->purchases as $purchase) {
            $sum += $purchase->totalPrice() * TaxesRules::getApplicableRate($this->country, $purchase->book);
        }
        return $sum;
    }

    public function getPurchasedBooks(): array
    {
        return $this->purchases;
    }


}

File: java/src/main/java/org/peaksys/gameoftechs/purchase/Invoice.java
package org.peaksys.gameoftechs.purchase;

import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.EqualsAndHashCode;
import lombok.ToString;
import org.peaksys.gameoftechs.IdGenerator;
import org.peaksys.gameoftechs.domain.country.Country;
import org.peaksys.gameoftechs.finance.TaxRule;

import java.beans.ConstructorProperties;
import java.util.ArrayList;
import java.util.List;

@ToString
@EqualsAndHashCode
public final class Invoice {
    @ToString.Include
    @EqualsAndHashCode.Include
    private final int id;
    @ToString.Include
    @EqualsAndHashCode.Include
    private final String clientName;
    @ToString.Include
    @EqualsAndHashCode.Include
    private final List<PurchasedBook> purchasedBooks;
    @ToString.Include
    @EqualsAndHashCode.Include
    private final Country country;

    public Invoice(String clientName, Country country) {
        this(IdGenerator.nextId(), clientName, country);
    }

    @ConstructorProperties({"id", "client", "country"})
    public Invoice(int id, String clientName, Country country) {
        this.id = id;
        this.clientName = clientName;
        this.country = country;
        this.purchasedBooks = new ArrayList<>();
    }

    public int getId() {
        return id;
    }

    public String getClientName() {
        return clientName;
    }

    public Country getCountry() {
        return country;
    }

    @JsonProperty("booksInBasket")
    public void addPurchasedBooks(List<PurchasedBook> books) {
        purchasedBooks.addAll(books);
    }

    public void addPurchasedBook(PurchasedBook book) {
        this.purchasedBooks.add(book);
    }

    public double computeTotalAmount() {
        return purchasedBooks.stream()
                .mapToDouble(purchasedBook -> purchasedBook.getTotalPrice() * TaxRule.getApplicableRate(country, purchasedBook.book()))
                .sum();
    }

    public List<PurchasedBook> getPurchasedBooks() {
        return purchasedBooks;
    }

}
